api/transport/response.go

```diff
@@ -68,3 +68,46 @@ type ResponseWriter interface {
 	// of Write().
 	SetApplicationError()
 }
+
+// ResponseMetaWriter returns a ResponseMeta struct that handlers and inbound
+// middleware may use to modify transport.Response fields.
+//
+// ResponseWriter implementations should implement this interface with the
+// intention of users attempting an upcast.
+type ResponseMetaWriter interface {
```

Comment:
We should elaborate on how we expect people to use this with an example.
Maybe change the second paragraph to,

```
// Middleware and transport implementations may attempt to upcast
// ResponseWriters to ResponseMetaWriters to access and write response
// metadata. Failure to cast MUST be handled.
//
//   if metaW, ok := resW.(transport.ResponseMetaWriter); ok {
//     meta := metaW.ResponseMeta()
//     meta.ID = "foo"
//   }
//
// Transport implementations that support writing response metadata should
// have their ResponseWriters implement ResponseMetaWriter to facilitate this.
```